### ISSUE TITLE
GitCommitBear: Require the commit message body contains a valid issue reference

### ISSUE THREAD

**Issue by jayvdb** [coala/coala-bears#1134]
Following on from https://github.com/coala/coala-bears/issues/1112 , some users may want the commit log to not contain a URL, but use the `#123` syntax.

Also, BitBucket [only supports](https://confluence.atlassian.com/bitbucket/resolve-issues-automatically-when-users-push-code-221451126.html) `#123` or `issue 123`, and doesnt support URLs.

**Comment by dracarys09** [comment]
@jayvdb I would like to work on this...

**Comment by dracarys09** [comment]
@jayvdb @Makman2 What we could do for this is give a setting say ```issue_reference_type``` values can be ```full_url``` or ```short_url``` and check accordingly. If short_url is given ```short_url``` can match ```#123``` type or ```issue 123``` type whereas if  full_url is given ```full_url``` will match full issue urls.
What are your thoughts on this ?

**Comment by jayvdb** [comment]
Imo this should be blocked on the other issue

**Comment by dracarys09** [comment]
@jayvdb which other issue are you referring to ?

**Comment by jayvdb** [comment]
See this issue description.

**Comment by dracarys09** [comment]
For now I'll find some other issue to work on then...

**Comment by Makman2** [comment]
@dracarys09 anyway the issue is blocked, I like the idea :+1: You should allow to combine those values, like
`issue_reference_type = short_url | full_url` or  (which is maybe better as we can reuse the coala `typed_list` type) `issue_reference_type = short_url, full_url` :)

**Comment by dracarys09** [comment]
@Makman2 yeah but some users would want to go for ```short_url``` type while others might prefer ```long_url``` type. By default we can check for the presence of any of these two types as you suggested if user does not give any of the settings otherwise we can check for the user defined setting.  😄

**Comment by Makman2** [comment]
yeah right haven't said something else :) But by default no check shall happen, so if a user wants to have at least a reference and he doesn't care whether it is a short or long url, we should allow to supply `issue_reference_type = short_url, long_url`, as omitting the setting completely means the user doesn't care whether any issue reference appears at all^^

**Comment by dracarys09** [comment]
yeah got your point...thanks for suggesting...hope that this issue gets unblocked soon...

**Comment by jayvdb** [comment]
Just remembered that loklak project requires a `Fixes #123` at the beginning of the shortlog.
See https://github.com/loklak/loklak_server/pull/866/commits .

**Comment by jayvdb** [comment]
Hmm, `Fixes #123` at the beginning of the shortlog can already be done with `shortlog_regex`.  The regex would need to contain the long list of available prefixes like `(Fixes|Closes|...)`, but often people only want a subset of the supported keywords, usually only `(Fixes|Closes)` which isnt too complicated.

**Comment by Makman2** [comment]
Hm maybe we can add convenience support, though I wouldn't concentrate on shortlogs yet^^

**Comment by dracarys09** [comment]
@jayvdb @Makman2  In ```check_issue``` method there is a parameter ```require_url``` and it's value can be ```short_ref```, ```long_ref``` which is already checking for short and full references. I think we should just add one more method say ```check_shortlog_issue_reference``` to check issue reference in shortlog. Rest we can do by making some modifications in ```check_issue``` method...

**Comment by Makman2** [comment]
right no extra setting, but the shortlog check should also happen in `check_issue`, otherwise stuff gets messy. In fact the issue check for the body could happen already in `check_body`, but it makes sense to keep `check_issue`^^

**Comment by dracarys09** [comment]
@Makman2  I have one more doubt... this might sound silly... since the changes should be made to ```check_issue``` but ```check_issue``` is written by @nkprince007 which is present at https://github.com/nkprince007/coala-bears/commit/8835efa3e36c081a3e59189ae8544fec81d76513

and this method is not yet merged in coala-bears repo... What should I do ?

**Comment by Makman2** [comment]
Don't rely on his commit and introduce `check_issue` on your own. Depending on which commit gets merged faster, the responsible person has to fix merge conflicts

**Comment by dracarys09** [comment]
But his commit has some code which I think is really useful. Shouldn't it be wrong if I directly use that code ?

**Comment by Makman2** [comment]
in this case you should rebase onto his commit. To do so you should checkout his branch via `git fetch origin pull/<ID-of-pull-request>/head:<some-branch-name>`, branch of your own and work on that. But remember that he might change his commits^^

**Comment by dracarys09** [comment]
@Makman2 I think his commit is in final stages as he just updated his PR with unit tests and all and I thing his commit will be merged in 3-4 days... Can I resume working on this after his commit gets merged and meanwhile start working on some other issue ?

**Comment by Makman2** [comment]
@dracarys09 sure :)

**Comment by jayvdb** [comment]
Note cross repo closes not yet supported is https://bitbucket.org/site/master/issues/15316/close-issues-with-a-commit-from-a-separate

c.f. https://github.com/coala/coala-bears/issues/2467 for another BitBucket commit format.
